src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.ProjectState.cs

```diff
@@ -369,26 +373,32 @@ public bool OnProjectRemoved(ProjectId id)
                 CancellationToken cancellationToken) where T : class
             {
                 var diagnostics = await DeserializeAsync(serializer, documentOrProject, key, stateKey, cancellationToken).ConfigureAwait(false);
-                if (diagnostics.IsDefault)
+                if (diagnostics == null)
                 {
                     return false;
                 }
 
-                add(key, diagnostics);
+                add(key, diagnostics.Value);
                 return true;
             }
 
-            private async Task<ImmutableArray<DiagnosticData>> DeserializeAsync(DiagnosticDataSerializer serializer, object documentOrProject, object key, string stateKey, CancellationToken cancellationToken)
+            private Task<StrongBox<ImmutableArray<DiagnosticData>>> DeserializeAsync(DiagnosticDataSerializer serializer, object documentOrProject, object key, string stateKey, CancellationToken cancellationToken)
             {
                 // check cache first
                 CacheEntry entry;
                 if (InMemoryStorage.TryGetValue(_owner.Analyzer, ValueTuple.Create(key, stateKey), out entry) && serializer.Version == entry.Version)
                 {
-                    return entry.Diagnostics;
+                    if (entry.Diagnostics.Length == 0)
+                    {
+                        // if there is no result, use cached task
+                        return s_emptyResultTaskCache;
```

Comment:
this particular method yes, I can do that, but I dont see the point since all the below one already return strong box one.